File: scripts/python/sacred_cli/card.py
"""Box-drawing card borders + word wrap. Mirrors scripts/cli/lib/card.js."""

from .ansi import COLORS, INVERSE, RESET, bg_hex, fg_hex, vis_len

# NOTE(jimmylee): Box-drawing characters — matches Card.tsx border rendering.
B = {"tl": "\u250C", "tr": "\u2510", "bl": "\u2514", "br": "\u2518", "h": "\u2500", "v": "\u2502"}

MIN_CARD_W = 10


def card_top(title: str, inner_w: int) -> str:
    t = f" {title} "
    fill = max(0, inner_w - 3 - len(t))
    return f"{B['tl']}{B['h']}{t}{B['h'] * fill}{B['tr']}"


def card_row(content: str, inner_w: int) -> str:
    inner = f"  {content}"
    vis = vis_len(inner)
    fill = inner_w - 2 - vis
    if fill >= 0:
        return f"{B['v']}{inner}{' ' * fill}{B['v']}"
    return f"{B['v']}{inner}{RESET}"


def card_header_row(content: str, inner_w: int) -> str:
    hdr_bg = bg_hex(COLORS["tableHeader"])
    hdr_fg = fg_hex(COLORS["text"])
    inner = f"  {content}"
    vis = vis_len(inner)
    fill = max(0, inner_w - 2 - vis)
    return f"{B['v']}{hdr_bg}{hdr_fg}{inner}{' ' * fill}{RESET}{B['v']}"


def card_bot(inner_w: int) -> str:
    return f"{B['bl']}{B['h'] * max(0, inner_w - 2)}{B['br']}"


def word_wrap(text: str, max_w: int):
    words = text.split(" ")
    lines = []
    cur = ""
    for word in words:
        if cur and len(cur) + 1 + len(word) > max_w:
            lines.append(cur)
            cur = word
        else:
            cur = cur + " " + word if cur else word
    if cur:
        lines.append(cur)
    return lines


def card_select_row(content: str, inner_w: int, selected: bool) -> str:
    if not selected:
        return card_row(content, inner_w)
    inner = f"  {content}"
    vis = vis_len(inner)
    fill = max(0, inner_w - 2 - vis)
    return f"{B['v']}{INVERSE}{inner}{' ' * fill}{RESET}{B['v']}"

File: scripts/cli/lib/card.ts
import { RESET, INVERSE, visLen, bgHex, fgHex, COLORS } from './ansi';

const B = { tl: '┌', tr: '┐', bl: '└', br: '┘', h: '─', v: '│' } as const;

const MIN_CARD_W: number = 10;

function cardTop(title: string, innerW: number): string {
  const t = ` ${title} `;
  const fill = Math.max(0, innerW - 3 - t.length);
  return `${B.tl}${B.h}${t}${B.h.repeat(fill)}${B.tr}`;
}

function cardRow(content: string, innerW: number): string {
  const inner = `  ${content}`;
  const vis = visLen(inner);
  const fill = innerW - 2 - vis;
  if (fill >= 0) return `${B.v}${inner}${' '.repeat(fill)}${B.v}`;
  return `${B.v}${inner}${RESET}`;
}

function cardHeaderRow(content: string, innerW: number): string {
  const hdrBg = bgHex(COLORS.tableHeader);
  const hdrFg = fgHex(COLORS.text);
  const inner = `  ${content}`;
  const vis = visLen(inner);
  const fill = Math.max(0, innerW - 2 - vis);
  return `${B.v}${hdrBg}${hdrFg}${inner}${' '.repeat(fill)}${RESET}${B.v}`;
}

function cardBot(innerW: number): string {
  return `${B.bl}${B.h.repeat(Math.max(0, innerW - 2))}${B.br}`;
}

function wordWrap(text: string, maxW: number): string[] {
  const words = text.split(' ');
  const lines: string[] = [];
  let cur = '';
  for (const word of words) {
    if (cur && cur.length + 1 + word.length > maxW) {
      lines.push(cur);
      cur = word;
    } else {
      cur = cur ? cur + ' ' + word : word;
    }
  }
  if (cur) lines.push(cur);
  return lines;
}

function cardSelectRow(content: string, innerW: number, selected: boolean): string {
  if (!selected) return cardRow(content, innerW);
  const inner = `  ${content}`;
  const vis = visLen(inner);
  const fill = Math.max(0, innerW - 2 - vis);
  return `${B.v}${INVERSE}${inner}${' '.repeat(fill)}${RESET}${B.v}`;
}

export { B, MIN_CARD_W, cardTop, cardRow, cardSelectRow, cardHeaderRow, cardBot, wordWrap };
